drivers/gpio/gpio_sim.c

```diff
@@ -0,0 +1,563 @@
+/*
+ * Copyright (c) 2020 Friedt Professional Engineering Services, Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT zephyr_sim_gpio
+
+#include <device.h>
+#include <drivers/gpio.h>
+#include <drivers/gpio/gpio_sim.h>
+#include <errno.h>
+#include <stdio.h>
+#include <zephyr.h>
+
+#include "gpio_utils.h"
+
+#define LOG_LEVEL CONFIG_GPIO_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(gpio_sim);
+
+#define interrupt_bitmask                                                      \
+	(0 | GPIO_INT_DISABLE | GPIO_INT_ENABLE | GPIO_INT_LEVELS_LOGICAL |    \
+	 GPIO_INT_EDGE | GPIO_INT_LOW_0 | GPIO_INT_HIGH_1)
+
+/*
+ * GPIO config / data
+ */
+
+struct gpio_sim_config {
+	struct gpio_driver_config common;
+	struct k_mutex mu;
+	sys_slist_t callbacks;
+	uint32_t flags[CONFIG_GPIO_SIM_NPINS_PER_PORT];
+	gpio_port_value_t vals;
+	gpio_port_pins_t interrupts;
+};
+
+struct gpio_sim_data {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+	const struct gpio_sim_config *config;
+};
+
+static struct gpio_sim_config gpio_sim_config[CONFIG_GPIO_SIM_NPORTS];
+static struct gpio_sim_data gpio_sim_data[CONFIG_GPIO_SIM_NPORTS];
+
+/*
+ * Util
+ */
+static gpio_port_pins_t
+gpio_sim_get_mask_of_pins_with_flags(struct gpio_sim_config *config,
+				     uint32_t flags)
+{
+	size_t i;
+	gpio_port_pins_t mask;
+
+	for (mask = 0, i = 0; i < CONFIG_GPIO_SIM_NPINS_PER_PORT; ++i) {
```

Comment:
Instead of the config, can you put the value in gpio_sim_data once (hopefully from DT) and use it from there?

Reply:
Good suggestion

Reply:
Done